Port various features from pbench-run-benchmark to pbench-linpack

ldoktor opened the pull request:
The pbench-run-benchmark uses pre/post scripts to generate the machine readable results for linpack, let's reuse that from the pbench-linpack binary.

webbnh reviewed (changes requested):
Hi Lukas, thanks for this PR -- there are some nice changes here.

I have no objections in principle to taking these changes.  Although, I will say, in the long term I think we would like to move away from these benchmark-specific scripts in favor of a common/external benchmark-runner.

With respect to your changes here, I have a few comments.
- [ ] We should discuss the feature you're adding to disable the RPM version check
- [ ] I don't think we want the "fake results" in the script itself -- those can be provided externally via mocking
- [ ] The "run directory" should not be hard-coded
- [ ] There's a mystery parameter to `generate-benchmark-summary`
- [ ] I don't think we want the "status" outputs (but I'll defer to Peter on that)
- [ ] There's housekeeping stuff:
  - use hard-tabs for indentation, one per level
  - use curly braces in shell variable references
  - we should have a little more discipline around variables with numeric values
  - it would be good to break up long lines

Also, there's a `cp` command inside the inner loop which probably deserves some examination.

ldoktor commented on agent/base:
```diff
@@ -160,6 +160,7 @@ function get_redhat_version() {
 }
 
 function check_install_rpm {
+    [ "$PBENCH_SKIP_CHECK_INSTALL_RPM" == true ] && return
```
Hello @webbnh, I'm actually struggling with `check_install_rpm` on many fronts. Previously I was playing with pbench on non-rpm-based systems, but I do understand that is not the primary usecase right now.

I also had to implement way to supply the correct version for fio and uperf on systems that comes with different version out of the box (I'm only comparing the same distros together so I don't really care about the different uperf version and besides I get notified of a different rpm version if that comparison happens). On older versions I'm also compiling fio myself (to get nbd backend), so the version logged by pbench is actually different than the one used via `benchmark_bin` (I'm treating this myself to get the used version instead).

As for linpack, it's even trickier as the hardcoded (not configurable) linpack version is `pbench-linpack-11.1.3.x86_64` and this version is even used for data-file location, while the only available version is an updated `pbench-linpack-11.1.3-1.x86_64`. Anyway the data-file location stayed the same (it's a minor update) so updating the version actually fails (because the data-file location based on rpm version differs) and relying on it also fails as it's not the hardcoded version.

For me I don't really need the rpm check as I just want to the version that is installed or specified via custom path. I'd welcome, though, to log the version based on the benchmark binary (`fio --version`).

portante replied:
> I also had to implement way to supply the correct version for fio and uperf on systems that comes with different version out of the box (I'm only comparing the same distros together so I don't really care about the different uperf version and besides I get notified of a different rpm version if that comparison happens). On older versions I'm also compiling fio myself (to get nbd backend), so the version logged by pbench is actually different than the one used via `benchmark_bin` (I'm treating this myself to get the used version instead).

@ldoktor, can you file an issue for this?

We don't have much support for non-RPM based linux distributions (see https://github.com/distributed-system-analysis/pbench/issues/91).  Any help towards that would be welcome.

As part of the above issue you open, we could do work to replace the RPM checks with simple path execution checks which are independent.

> As for linpack, it's even trickier as the hardcoded (not configurable) linpack version is `pbench-linpack-11.1.3.x86_64` and this version is even used for data-file location, while the only available version is an updated `pbench-linpack-11.1.3-1.x86_64`. Anyway the data-file location stayed the same (it's a minor update) so updating the version actually fails (because the data-file location based on rpm version differs) and relying on it also fails as it's not the hardcoded version.

Let's take the step of changing `pbench-linpack` to not hard-code the version and move it to the `pbench-agent-default.cfg` file, with a minimum supported version, like we did for `fio`, `uperf`, etc.

> 
> For me I don't really need the rpm check as I just want to the version that is installed or specified via custom path. I'd welcome, though, to log the version based on the benchmark binary (`fio --version`).

Tracking that request in an issue so that we could schedule the work, or accept a community contribution, would be great.

portante replied:
Regardless, I'd like to avoid this kind of change in this PR and focus on the functional enhancements ported over from pbench-run-benchmark.

ldoktor replied:
OK, let me re-send this separately. Anyway let me keep it here as well as without it there is no way to check this (the pbench-linpack of the desired version does not exists)

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -45,14 +45,15 @@ sysinfo="default"
 function usage {
 	printf "\tThe following options are available:\n\n"
 	printf -- "\t-C str --config=str         name of the test config\n"
-	printf -- "\t       --threads=int[,int]  number of threads to use (default is num_cpus)\n"
+    printf -- "\t       --samples=<int>      number of samples to use per test iteration (default is $nr_samples)\n"
```
Yep, sorry for that, I mean to fix that before the final version (for other projects I'm using my vim preset with softtabs; the good old spaces vs. tabs issue...)

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -99,6 +107,10 @@ while true; do
 		shift;
 		break;
 		;;
+        --xxx-fake-results)
+        xxx_fake_results=true
+        shift;
+        ;;
```
I understand, this was the fastest way I can develop things. I haven't found a documentation about the selftests and without it it looked like too complex to try to understand it for the sake of contributing this.

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+    done
+    let count=$count+1 # now we can move to the next iteration
+done
+echo "Postprocessing all iterations"
+postprocess-results "$benchmark_run_dir"
 
 # Stop the tool meisters on each registered local/remote host
+echo "Stopping tool-meister"
```
sure

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
-pbench-postprocess-tools --group=$tool_group --dir=$benchmark_results_dir
-
-ln -s ${sample} $benchmark_run_dir/$iteration/reference-result
+count=1
```
ok

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+        pbench-postprocess-tools --group=$tool_group --dir=$benchmark_results_dir
+        postprocess-iteration "$benchmark_results_dir" "$benchmark_run_dir"
+    done
+    let count=$count+1 # now we can move to the next iteration
```
I wanted to keep the current code-style, it's a copy&paste from pbench-fio... anyway I do prefer the ++ operator so let me update that in the next version :-)

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+for thread in ${threads//,/ }; do
+    iteration="$count-$thread-threads"
+    echo $iteration >> $benchmark_iterations
+    for sample in `seq 1 $nr_samples`; do
```
Again, copy&paste from pbench-fio, I'll gladly update that :-)

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+    echo "$cmd" > "$filename"
+    chmod +x "$filename"
+    ./$filename &> $filename.out
+    ret="$?"
```
ack

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+    ./$filename &> $filename.out
+    ret="$?"
+    popd >/dev/null
+    if [ "$ret" -ne 0 ]; then
```
ack

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+
+function postprocess-iteration {
+    echo "Postprocessing iteration data"
+    store_and_run "$1" "'$pbench_bin/bench-scripts/postprocess/linpack-postprocess-cdm' '$1' '$2' '$tool_group' 1 html" "$benchmark-postprocess.cmd"
```
ack

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+    if [ "$ret" -ne 0 ]; then
+        debug_log "failed to execute: $workdir/$filename)"
+    fi
+    return "$ret"
```
ack

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+function postprocess-iteration {
+    echo "Postprocessing iteration data"
+    store_and_run "$1" "'$pbench_bin/bench-scripts/postprocess/linpack-postprocess-cdm' '$1' '$2' '$tool_group' 1 html" "$benchmark-postprocess.cmd"
+    return "$?"
```
now when you say that I remember using that long time ago... long time without bash...

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+
+function postprocess-results {
+    echo "Postprocessing all iterations data"
+    store_and_run "$1" "'$pbench_bin/bench-scripts/postprocess/generate-benchmark-summary' '$benchmark' '$orig_cmd' '$1'" "generate-benchmark-summary.cmd"
```
orig_cmd is set at the beginning of the pbench-linpack script, it stores the original argvs

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+        benchmark_results_dir="$benchmark_run_dir/$iteration/sample$sample"
+        mkdir -p $benchmark_results_dir
+
+        # Generate the linpack.input dirs (in /var/lib/pbench-agent/tmp)
+        store_and_run "$benchmark_results_dir" "'$pbench_bin/bench-scripts/postprocess/linpack-prepare-input-file' --threads '$thread' --run-samples 1 --linpack-binary '$linpack_dir/$linpack_cmd'" "$benchmark-preprocess.cmd"
+        mkdir -p $benchmark_results_dir/clients/localhost
+        cp /var/lib/pbench-agent/tmp/linpack* "$benchmark_results_dir/clients/localhost"
```
this is actually hardcoded in the `linpack-prepare-input-file`, which I didn't wanted to change to not to interfere with the fragile ansible-based execution.

webbnh replied:
See [below](https://github.com/distributed-system-analysis/pbench/pull/2710#discussion_r838962658).

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+        # Generate the linpack.input dirs (in /var/lib/pbench-agent/tmp)
+        store_and_run "$benchmark_results_dir" "'$pbench_bin/bench-scripts/postprocess/linpack-prepare-input-file' --threads '$thread' --run-samples 1 --linpack-binary '$linpack_dir/$linpack_cmd'" "$benchmark-preprocess.cmd"
+        mkdir -p $benchmark_results_dir/clients/localhost
+        cp /var/lib/pbench-agent/tmp/linpack* "$benchmark_results_dir/clients/localhost"
```
The main issue here is the `linpack-prepare-input-file` script, which uses hardcoded location and I was hesitant to rely on it to be kept alike after the iteration execution (one never knows when some script updates this temporary location), so I deliberately kept executing it for each sample. Anyway I can move it ouside the `sample` loop and execute it only inside the `threads` loop (it generates the file based on the `threads` value). Please let me know whether you really want me to do that as I don't generally trust hardcoded tmp locations to staly up-to-date.

webbnh replied:
My inclination at the moment is to suggest that you modify `linpack-prepare-input-file` to allow the value of its `output_dir` to be overridden by a (new) command line switch.  I think that that is a simple, targeted, compatible change, which will allow `pbench-linpack` to direct it to place the output files directly into the right spot, without having to copy them here.  And then, you can replace uses of `/var/lib/pbench-agent/tmp` here with `$pbench_tmp` without worrying about a mismatch between this script and the other one.

(`linpack-prepare-input-file` is already copying the Linpack binary to the output directory; my suggested change will avoid copying it a second time on each iteration....)

portante commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+        mkdir -p $benchmark_results_dir
+
+        # Generate the linpack.input dirs (in /var/lib/pbench-agent/tmp)
+        store_and_run "$benchmark_results_dir" "'$pbench_bin/bench-scripts/postprocess/linpack-prepare-input-file' --threads '$thread' --run-samples 1 --linpack-binary '$linpack_dir/$linpack_cmd'" "$benchmark-preprocess.cmd"
```
Seems like `linpack-prepare-input-file` needs to be added to this commit?

webbnh replied:
`agent/bench-scripts/postprocess/linpack-prepare-input-file` is an existing file in the Pbench repo...are you expecting changes to it?

portante replied:
I was expecting it to be renamed to what the other bench-scripts do, but using it as it stands is okay.

portante commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+        pbench-stop-tools --group=$tool_group --dir=$benchmark_results_dir
+        pbench-send-tools --group=$tool_group --dir=$benchmark_results_dir
+        pbench-postprocess-tools --group=$tool_group --dir=$benchmark_results_dir
+        postprocess-iteration "$benchmark_results_dir" "$benchmark_run_dir"
```
I think we want to follow the same convention from other bench-scripts here, see https://github.com/distributed-system-analysis/pbench/blob/main/agent/bench-scripts/pbench-uperf#L772.

webbnh replied:
I _think_ Lukas's code _is_ doing that...he cleverly wrapped it up inside a couple of utility functions.

ldoktor replied:
Well I dislike repeating my code so I used the `store_and_run` function. I think it'd actually make sense to move it to `base` but I'm too fresh to this project to consider something like that :-) Anyway is the current approach fine or do you want me to adjust it somehow?

webbnh replied:
I, too, think making it generally available would be good.  (It would be nice to get all of the benchmark scripts consistent, where we can.)

portante replied:
... in another PR.  =)

portante commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,48 +120,131 @@ while true; do
+    let count=$count+1 # now we can move to the next iteration
+done
+echo "Postprocessing all iterations"
+postprocess-results "$benchmark_run_dir"
```
And `process-iteration-samples` like https://github.com/distributed-system-analysis/pbench/blob/main/agent/bench-scripts/pbench-uperf#L776.

webbnh replied:
Same [answer](https://github.com/distributed-system-analysis/pbench/pull/2710#discussion_r836874834) but different.

portante replied:
Okay, caught by the different conventions.

portante reviewed (changes requested):
Thanks @ldoktor, let me know how I can help make this move along.

ldoktor commented:
Hello folks, addressed hopefully all required changes, removed the `PBENCH_SKIP_CHECK_INSTALL_RPM` commit as well as the testing "dry-run" commit.

I have tried moving from rpm check to `linpack -v` kind of check but unfortunately linpack does not provide such information (only -h and -e but none contains the version). For testing purposes I removed the rpm check manually as I don't poses of the right linpack rpm version so having a global switch would really help me.

ldoktor commented:
@portante as for the selftests, I changed the parameters to show the actual numbers instead of the human readable output (16 vs. num_cpus), do you have a way to mask those in selftests?

portante reviewed (changes requested):
Looks good, but let's get @webbnh's comments addressed.

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,27 +116,45 @@ while true; do
+	ret=$?
+	popd >/dev/null
+	if [ $ret -ne 0 ]; then
+		debug_log "failed to execute: $workdir/$filename)"
```
I'm sorry it's a remnant from the original `($value)` version but I dropped the parenthesis. Let me remove that

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,27 +116,45 @@ while true; do
+	pushd "$1" >/dev/null
+	echo "$cmd" > "$filename"
+	chmod +x "$filename"
+	./$filename &> $filename.out
+	ret=$?
+	popd >/dev/null
+	if [ $ret -ne 0 ]; then
+		debug_log "failed to execute: $workdir/$filename)"
+	fi
+}
```
sure

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -138,16 +164,37 @@ fi
+
+		# Generate the linpack.input dirs (in /var/lib/pbench-agent/tmp)
+		store_and_run "$benchmark_results_dir" "'$pbench_bin/bench-scripts/postprocess/linpack-prepare-input-file' --threads '$thread' --run-samples 1 --linpack-binary '$linpack_dir/$linpack_cmd'" "$benchmark-preprocess.cmd"
+		mkdir -p $benchmark_results_dir/clients/localhost
+		cp /var/lib/pbench-agent/tmp/linpack* "$benchmark_results_dir/clients/localhost"
```
ack the lines 175 and 179, but I'm not sure where would you like to put the parenthesis on the line 180.

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -138,16 +164,37 @@ fi
+		## Run the benchmark and start/stop perf analysis tools
+		pbench-start-tools --group=$tool_group --dir=$benchmark_results_dir
+
+		pushd "$benchmark_results_dir/clients/localhost/" >/dev/null
+		./linpack.sh
+		popd >/dev/null
```
Nice, haven't used that construct for a while :-) Let me use that wherever it makes sense.

ldoktor commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -108,27 +116,45 @@ while true; do
+	filename="$3"
+	pushd "$1" >/dev/null
+	echo "$cmd" > "$filename"
+	chmod +x "$filename"
+	./$filename &> $filename.out
```
ack

ldoktor commented:
Addressed hopefully all issues including the pushd/popd replacement with a subshells and the linpack run script generation once per iteration, rather than once per sample.

webbnh commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -138,16 +166,36 @@ fi
+		## Run the benchmark and start/stop perf analysis tools
+		pbench-start-tools --group=$tool_group --dir="$sample_dir"
+
+		( cd "$iteration_dir" && ./linpack.sh )
+		mv "$iteration_dir/"linpack.{out,meta} "$sample_dir/clients/localhost"
+
+		pbench-stop-tools --group=$tool_group --dir="$sample_dir"
+		pbench-send-tools --group=$tool_group --dir="$sample_dir"
+		pbench-postprocess-tools --group=$tool_group --dir="$sample_dir"
+		postprocess-iteration "$sample_dir" "$benchmark_run_dir"
```
It's a small detail, but shouldn't we be stopping the tools before moving the output files?  (I realize that moving the files is probably a cheap operation, but we don't want that operation included in the data which the tools are collecting, right?)

ldoktor replied:
Yes, you are right, that's why reviews are so important :-)

webbnh reviewed (dismissed):
It looks very good, Lukas.  Thanks for all the effort!

portante reviewed (changes requested):
Looks good, just need to stop the tools before moving the directories around.  Nice catch, @webbnh!

webbnh commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -138,16 +166,37 @@ fi
+		pbench-stop-tools --group=$tool_group --dir="$sample_dir"
+		pbench-send-tools --group=$tool_group --dir="$sample_dir"
+		# Move the results to the output directory expected by the postprocess script
+		mv "$iteration_dir/"linpack.{out,meta} "$benchmark_results_dir/clients/localhost"
```
Where is `$benchmark_results_dir` defined?  (And, what is it defined to?)

Previously, it was defined to `"$benchmark_run_dir/$iteration/${sample}"`.  I think this would now be equivalent to `"$sample_dir"` (which is supported by the various references to it in lines 185-194).

So, it looks like line 192 should be
```
		mv "$iteration_dir/"linpack.{out,meta} "$sample_dir/clients/localhost"
```
which would match line 182.

Am I missing something?

ldoktor replied:
Oups, my bad. I forgot to change this in the `Unify output dir naming in pbench-linpack` commit. Addressed now.

ldoktor replied:
(I only tested it before renaming the variables, it was just a mass rename, what could happen, right?)

webbnh reviewed (approved):
Hi @ldoktor, this looks good...except possibly for one other item.  Could you confirm?

portante commented:
@ldoktor, @webbnh, please consider merging https://github.com/ldoktor/pbench/pull/4 into this PR.

portante reviewed (changes requested):
We need tests provided in https://github.com/ldoktor/pbench/pull/4.

ldoktor commented:
Thank you, Peter, for adding the selftests. I tested it functionally, reviewed your code and looks good to me.

webbnh commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -34,101 +24,167 @@ threads=`cat /proc/cpuinfo | grep processor | wc -l`
+	local cmd=${2}
+	local filename=${3}
+	(
+		cd "${1}" &&
```
Shouldn't that be
```
		cd "${workdir}" &&
```
🙃

portante replied:
Fixed in https://github.com/ldoktor/pbench/pull/5.

webbnh commented on agent/bench-scripts/pbench-linpack:
```diff
@@ -138,21 +194,52 @@ fi
+let count=1
+for thread in ${threads//,/ }; do
+	iteration="${count}-${thread}-threads"
+	echo ${iteration} >> ${benchmark_iterations}
```
This is done inside `record_iteration`...are we sure we want to do it here as well?

portante replied:
Fixed in https://github.com/ldoktor/pbench/pull/5.

webbnh commented on agent/bench-scripts/tests/pbench-linpack/test-65.pre:
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+# Setup a fake install of linpack but without the expected executable.
```
s/without/with a stub for/ ?

portante replied:
Fixed in https://github.com/ldoktor/pbench/pull/5.

webbnh reviewed (approved):
Looks OK, but we should probably fix the items I pointed out.

ldoktor commented:
Thank you, guys, for taking care of the polishing.
